sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala

```diff
@@ -60,6 +70,45 @@ object Literal {
   }
 
   /**
+   * Returns the Spark SQL DataType for a given class object. Since this type needs to be resolved
+   * in runtime, we use match-case idioms for class objects here. However, there are similar
+   * functions in other files (e.g., HiveInspectors), so these functions need to merged into one.
+   */
+  private[this] def componentTypeToDataType(clz: Class[_]): DataType = clz match {
+    // primitive types
+    case c: Class[_] if c == jShort.TYPE => ShortType
```

Comment:
Here is an example:
```scala
  case object container {def value: Int = 42}
  case object Container {def value: Int = 42}

  def extract(wrapper: Any) = wrapper match {
    case container => container.value // error: value "value" is not a member of Any
    case Container => Container.value // ok
  }
```
It's not a very good example, maybe something like http://stackoverflow.com/a/6576663/917519 will give you a better use-case